src/charm_state.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Module that handles Pollen charm's state."""
+
+from typing import TypedDict
+
+from pydantic import BaseModel, Extra
+
+
+class WebsiteDict(TypedDict):
+    """Custom dict for the website property.
+
+    Attrs:
+        hostname: Website hostname.
+        port: Website port.
+    """
+
+    hostname: str
+    port: str
+
+
+class WebsiteModel(BaseModel, extra=Extra.forbid):
+    """Pydantic validation model for the website property.
+
+    Attrs:
+        hostname: Website hostname.
+        port: Website port.
+    """
+
+    hostname: str
+    port: str
+
+
+class CharmState:
+    """Class that represents the state of the Pollen charm.
+
+    Attrs:
+        website: Website to pass over the website relation
+    """
+
+    def __init__(self, hostname: str | None) -> None:
+        """Construct.
+
+        Args:
+            hostname: hostname to build the website property
+        """
+        self._hostname = hostname
+
+    @property
+    def website(self) -> WebsiteDict:
+        """The state for this application of the website.
+
+        Returns: a WebsiteDict object to be used by the website relation.
+        """
+        website_data = {"hostname": format(self._hostname), "port": "443"}
+        website_model = WebsiteModel(**website_data)
+        return website_model
```

Comment:
That's also typically some code to cover with unittest and some happy/non-happy cases

I'm not sure you necessarily need to model WebsiteDict just return a dict, it will avoid having some repeated attributes between pydantic model and the typeddict

See flask_config in the Flask charm: https://github.com/canonical/flask-k8s-operator/blob/main/src/charm_state.py

Last comment, I think the 443 value should be a constant, will be easier to use in integration tests and avoid hardcoded magic numbers